modules/core/src/main/scala/gem/Location.scala

```diff
@@ -10,46 +11,167 @@ import scalaz.Ordering.{EQ, GT, LT}
   * `Location` is a proper prefex of another, then it sorts ahead of the other
   * `Location`.
   */
-sealed abstract case class Location(toIList: IList[Int]) {
+sealed trait Location {
 
-  // Use the sketchy sealed abstract case class technique to control the
-  // NonEmptyList values that are passed to the constructor.  We want to
-  // ensure that trailing Int.MinValue is always trimmed so that EQ and ==
-  // agree.
+  // These functions aren't of any use to clients.  Instead they are involved
+  // in the cacluation of Locations that fall between two other locations.
 
-  def toList: List[Int] =
-    toIList.toList
+  /** Infinite Stream of position elements corresponding to this Location. */
+  protected def positions: Stream[Int]
 
-  override def toString: String =
-    toList.mkString("{", ",", "}")
+  /** Minimum size required to obtain all the fixed elements in this Location,
+    * if any.  Beginning and End Locations have no fixed elements.
+    */
+  protected def minPrefixLength: Int
 }
 
 object Location {
-  private def trim(is: Seq[Int]): IList[Int] =
-    trim(IList(is: _*))
 
-  // Remove trailing Int.MinValue so that == agrees with EQ.
-  private def trim(is: IList[Int]): IList[Int] =
-    is.dropRightWhile(_ == Int.MinValue)
+  /** A marker for the first Location.  No other Location comes before the
+    * Beginning.  Beginning should not be used to order elements unless you
+    * are sure no elements will ever need to be inserted before Beginning.  Use
+    * Beginning to calculate Locations that fall before the first existing
+    * Location (if any).
+    */
+  case object Beginning extends Location {
+    protected def positions: Stream[Int] =
+      Stream.continually(Int.MinValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{α}"
+  }
+
+  /** A Location that falls somewhere between the Beginning and End.
+    */
+  sealed abstract case class Middle(posList: OneAnd[IList, Int]) extends Location {
+    def positions: Stream[Int] =
+      posList.toStream #::: Stream.continually(Int.MinValue)
+
+    def toIList: IList[Int] =
+      posList.head +: posList.tail
+
+    def toList: List[Int] =
+      posList.head :: posList.tail.toList
+
+    protected def minPrefixLength: Int =
+      1 + posList.tail.length
```

Comment:
We might consider `OneAnd[Vector, Int]` here if this ends up being a performance issue. Looks like it's only used to compute new locations for insert though so maybe it doesn't matter.